tstest/clock.go

```diff
@@ -4,57 +4,495 @@
-	// Start is the first value returned by Now.
-	Start time.Time
-	// Step is how much to advance with each Now call.
-	Step time.Duration
-	// Present is the time that the next Now call will receive.
-	Present time.Time
+	// start is the first value returned by Now. It must not be modified after
+	// init() is called.
+	start time.Time
+
+	mu sync.Mutex
 
-	sync.Mutex
+	// step is how much to advance with each Now call.
+	step time.Duration
+	// present is the last value returned by Now (and will be returned again by
+	// ConstantNow).
+	present time.Time
+	// skipStep indicates that the next call to Now should not add step to
+	// present. This occurs after initialization and after Advance.
+	skipStep bool
+	// timerChannelSize is the buffer size to use for channels created by
+	// NewTimer and NewTicker.
+	timerChannelSize int
+
+	events eventManager
+}
+
+func (c *Clock) init() {
+	if c.start.IsZero() {
+		c.start = time.Now()
+	}
+	if c.timerChannelSize == 0 {
+		c.timerChannelSize = 1
+	}
+	c.present = c.start
+	c.skipStep = true
+	c.events.AdvanceTo(c.present)
 }
 
 // Now returns the virtual clock's current time, and advances it
 // according to its step configuration.
 func (c *Clock) Now() time.Time {
-	c.Lock()
-	defer c.Unlock()
-	c.initLocked()
-	step := c.Step
-	ret := c.Present
-	c.Present = c.Present.Add(step)
-	return ret
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.skipStep && c.step > 0 {
+		c.present = c.present.Add(c.step)
+		c.events.AdvanceTo(c.present)
+	}
+	c.skipStep = false
+	return c.present
+}
+
+// ConstantNow returns the virtual clock's current time but does not advance it
+// if a step is configured.
+func (c *Clock) ConstantNow() time.Time {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.present
+}
+
+func (c *Clock) Advance(d time.Duration) time.Time {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.skipStep = true
+
+	if d > 0 {
+		c.present = c.present.Add(d)
+		c.events.AdvanceTo(c.present)
+	}
+
+	return c.present
+}
+
+func (c *Clock) Start() time.Time {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.start
+}
+
+func (c *Clock) Step() time.Duration {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.step
+}
+
+func (c *Clock) SetStep(d time.Duration) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.step = d
 }
 
-func (c *Clock) Advance(d time.Duration) {
-	c.Lock()
-	defer c.Unlock()
-	c.initLocked()
-	c.Present = c.Present.Add(d)
+func (c *Clock) SetTimerChannelSize(n int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.timerChannelSize = n
 }
 
-func (c *Clock) initLocked() {
-	if c.Start.IsZero() {
-		c.Start = time.Now()
+// NewTicker returns a Ticker that uses this Clock for accessing the current
+// time.
+func (c *Clock) NewTicker(d time.Duration) (tstime.TickerController, <-chan time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	t := &Ticker{
+		nextTrigger: c.present.Add(d),
+		period:      d,
+		em:          &c.events,
 	}
-	if c.Present.Before(c.Start) {
-		c.Present = c.Start
+	t.init(c.timerChannelSize)
```

Comment:
hmm, so that's what timerChannelSize is for. Isn't this channel always size 1 in the real ticker/timer implementations? With specified semantics for what happens if events occur faster than they're processed?

I lack context here, so maybe I'm way off base. At a glance, this feels like it shouldn't be configurable, because it lets you create tickers and timers with subtly different semantics to the production ones?

Reply:
Yes it's always 1 in real timers/tickers. The main place it comes up is if you have a 1 second ticker and jump time forward by a minute, now there are 60 ticks that appear all at once with no chance to have a draining goroutine get scheduled and do something meaningful. On the other hand, in other situations maybe you want to test what happens if a tick is lost (which is annoyingly difficult to create a test for without being able to arrange for that) so I feel like it needs to be an option.